provisioner/template.go

```diff
@@ -733,3 +743,55 @@ func sumQuantities(quantities ...string) (string, error) {
 func awsValidID(id string) string {
 	return strings.Replace(id, ":", "__", -1)
 }
+
+// instanceTypeCPU returns the vCPUs of an instance type provided as k8sresource.Quantity represented as string
+func instanceTypeCPU(context *templateContext, instanceType string) (string, error) {
+	// get the instance type info
+	instanceTypeInfo, err := context.instanceTypes.InstanceInfo(instanceType)
+
+	if err != nil {
+		return "", err
+	}
+
+	cpu := fmt.Sprintf("%v", instanceTypeInfo.VCPU)
+
+	return cpu, nil
+}
+
+// instanceTypeMemory returns the memory of an instance type provided as k8sresource.Quantity represented as string
+func instanceTypeMemory(context *templateContext, instanceType string) (string, error) {
+	// get the instance type info
+	instanceTypeInfo, err := context.instanceTypes.InstanceInfo(instanceType)
+
+	if err != nil {
+		return "", err
+	}
+
+	memory := k8sresource.NewQuantity(instanceTypeInfo.Memory, k8sresource.BinarySI)
+
+	return memory.String(), nil
+}
+
+// scaleQuantity scales a k8sresource.Quantity by a factor, represented as string
+// returns the k8sresource.Quantity and an error if the scaling factor is less than or equal to 0.0
+func scaleQuantity(quantityStr string, factor float32) (string, error) {
+	// validate scaling factor
+	if factor <= 0.0 {
+		return quantityStr, fmt.Errorf("scaling factor must be greater than 0.0")
+	}
+
+	// parse the quantity as k8sresource.Quantity
+	quantity, err := k8sresource.ParseQuantity(quantityStr)
+	if err != nil {
+		return quantityStr, fmt.Errorf("failed to parse %v as k8sresource.Quantity: %v", quantityStr, err)
+	}
+	// CPU quantities can be represented as milli-cores, so we need to scale them differently
+	if quantity.Value() < 1024 && quantity.Format == k8sresource.DecimalSI {
+		quantity.SetMilli(int64(float32(quantity.MilliValue()) * factor))
+	} else {
+		// memory quantities can be represented as binary SI, so we need to scale them differently
+		scaledValue := int64(float32(quantity.Value()) * factor)
```

Comment:
Maybe https://pkg.go.dev/k8s.io/apimachinery/pkg/api/resource#Quantity.AsApproximateFloat64 fits better here.

quantity.Value() rounds to integer, e.g. I am not sure if it will do the right thing for `scaleQuantity "2.5Gi", 1`- it should retrurn 2.5Gi (maybe as bytes) but I think it may round 2.5 to 2 first.

Reply:
Let's add a test for this @demonCoder95

Reply:
That's a good catch. It in fact returns `2560Mi` instead of `2.5Gi`. It didn't at least round it to `2Gi`. I'm thinking when `Gi` units become fractions, it reduces the answer to `Mi`. Is that okay?

Reply:
It makes sense to show `Mi` instead of `Gi`

Reply:
Using `AsApproximateFloat64` didn't change this result. The result is the same as I described above. There is no rounding off in terms of `Gi` units but some rounding errors when going from `Gi` -> `Mi` units.

Reply:
Actually, nevermind, 2.5Gi -> 2048Mi + 512Mi -> 2560Mi. So, it's doing correctly 😁  I think it's good! I added the test case in the next commit.

Reply:
BTW, I'd suggest to always prefer float64 unless its really math performance-critical part (i.e. never in our domain).
Float numbers represent integers in certain range without loss of precision, for float32 the limit is quite low, see https://stackoverflow.com/a/3793950/828328